### DISCUSSION TITLE
Task Implementation

### DISCUSSION THREAD

**Discussion by AndrePointner** [Ulm-IQO/qudi-core#105] [General]
Hi,

I was wondering what the best practice is to implement automated measurements in the current Qudi-Core iteration. I couldn't find a working documentation on the TaskRunner Module and from what I've read from the old versions documentation, the TaskRunner is meant more for automating minuscule tasks such as keeping an optimizer running in regular intervals.

Would it be feasible to implement e.g. a ODMR Power Sweep as a Task (and if so: how?) or is a Script Implementation with Remote Modules more in line with the idea of Qudi?

Thanks!

**Comment by timoML** [comment]
In principle the TaskRunner could be used for such automization.
However, it is only providing a relatively basic state machine and one would need to add quite some code for control flow logic of experiments.

The way we do it here in Ulm is via jupyter notebooks in conjunction with script libraries.
1. In jupyter, a list of dicts is created. Each dict contains the information about experiments to be performed.
2. The library script takes these lists and starts the specified experiment.
3. Then, the script runs an execution loop. This loop is periodically checking, whether the conditions (e.g. measurement time or some flags) are met to continue with the next experiment in the script.

Unfortunately, we don't currently have the scripts at a code quality level where we would want to distribute them.
If you had the time to design and implement a proper solution, we can certainly assist you. I think it would be valueable to a lot of people.

**Reply by AndrePointner** [reply]
Thanks for the Feedback. This is similar to the solution I implemented. I just use rpyc and remote modules directly, because I'm irrationally alergic to Jupyter.

One more thing on the topic: Is it safer to remotely access the GUI and benefit from built in thread safety methods or should I just access the underlying logic directly when scripting measurements? Whats the best practice in this regard?

Thanks!

**Reply by Neverhorst** [reply]
Certainly using the logic directly. 
The thread-safety (if any) is built into the logic modules public methods and properties (non-underscore names). 
Using the GUI directly may sometimes work but in general it is more likely to cause all kinds of problems.
